### PR TITLE
Add BMP parser

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Michal Zientkiewicz <mzient@nvidia.com>
Co-authored-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Adds a new parser to the imgcodec module, ported from the previous implementation in dali/image.
- The new parse implements the ImageParser interface from the imgcodec module
- Allows to get image info from BMP headers.

### Affected modules and functionalities:

No code affected

### Key points relevant for the review:

NA

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/parsers/bmp.cc [hunk 1]
```diff
@@ -12,19 +12,150 @@
...
+  int32_t header_size;
+  int32_t width, heigth;
+  uint16_t planes, bpp;
+  uint32_t compression, image_size;
+  int32_t x_pixels_per_meter, y_pixels_per_meter;
+  uint32_t colors_used, colors_important;
+};
+static_assert(sizeof(BitmapInfoHeader) == 40);
+
 ImageInfo BmpParser::GetInfo(ImageSource *encoded) const {
```
Comment (stiepan) [inline dali/imgcodec/parsers/bmp.cc:88]: Could you maybe add a link to the BMP header description as it was the case for the existing parsing funcionality?

File: dali/imgcodec/parsers/bmp.cc [hunk 2]
```diff
@@ -12,19 +12,150 @@
...
+    h = abs(header.heigth);
+    bpp = header.bpp;
+    compression_type = header.compression;
+    ncolors = header.colors_used;
+    if (bpp <= 8) {
+      palette_start = stream->TellRead();
+      palette_entry_size = 4;
+      ncolors = ncolors == 0 ? 1_uz << bpp : ncolors;
+    }
+  }
```
Comment (stiepan) [inline dali/imgcodec/parsers/bmp.cc:128]: What if none of the two branches is taken? Are fine with calling number_of_channels (I guess) and returning zeros?

**Review by stiepan — APPROVED** [review]

File: dali/imgcodec/parsers/bmp.cc (hunk 2 above)
Comment (jantonguirao) [reply dali/imgcodec/parsers/bmp.cc:128]: Looking at the info here: https://en.wikipedia.org/wiki/BMP_file_format#DIB_header_(bitmap_information_header)
I understand the following:
- We support BITMAPCOREHEADER (12 bytes long)
- We support BITMAPINFOHEADER and all following versions (40, 52, 56, 108, 124 bytes long respectively)
Given that, we should not fall outside of those two cases. If that happens, the arguments of number_of_channels will be 0 which will result in an exception thrown.

@mzient can you double check my understanding is correct?

File: dali/imgcodec/parsers/bmp.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/imgcodec/parsers/bmp.cc:88]: Will do

File: dali/imgcodec/parsers/bmp.cc (hunk 2 above)
Comment (mzient) [reply dali/imgcodec/parsers/bmp.cc:129]: If none of the two branches is taken then the file is malformed and parsing should fail - preferably here and not somewhere later.

Comment (mzient) [reply dali/imgcodec/parsers/bmp.cc:147]: I've added an error.

**Pull request merged**